src/visitor/DeduceTypeVisitor.cpp

```diff
@@ -0,0 +1,408 @@
+    DCHECK(qctx != nullptr);
+    DCHECK(vctx != nullptr);
+}
+
+void DeduceTypeVisitor::visit(ConstantExpression *expr) {
+    QueryExpressionContext ctx(nullptr);
+    type_ = expr->eval(ctx(nullptr)).type();
+}
+
+void DeduceTypeVisitor::visit(UnaryExpression *expr) {
+    expr->operand()->accept(this);
+    if (!ok()) return;
+    switch (expr->kind()) {
+        case Expression::Kind::kUnaryPlus:
+            break;
+        case Expression::Kind::kUnaryNegate: {
+            DETECT_UNARYEXPR_TYPE(-);
+            break;
+        }
+        case Expression::Kind::kUnaryNot: {
+            DETECT_UNARYEXPR_TYPE(!);
+            break;
+        }
+        case Expression::Kind::kUnaryIncr: {
+            auto detectVal = kConstantValues.at(type_) + 1;
+            if (detectVal.isBadNull()) {
+                std::stringstream ss;
+                ss << "`" << expr->toString() << "' is not a valid expression, "
+                   << "can not apply `++' to " << type_ << ".";
+                status_ = Status::SemanticError(ss.str());
+                return;
+            }
+            type_ = detectVal.type();
+            break;
+        }
+        case Expression::Kind::kUnaryDecr: {
+            auto detectVal = kConstantValues.at(type_) - 1;
+            if (detectVal.isBadNull()) {
+                std::stringstream ss;
+                ss << "`" << expr->toString() << "' is not a valid expression, "
+                   << "can not apply `--' to " << type_ << ".";
+                status_ = Status::SemanticError(ss.str());
+                return;
+            }
+            type_ = detectVal.type();
+            break;
+        }
+        default: {
+            LOG(FATAL) << "Invalid unary expression kind: " << static_cast<uint8_t>(expr->kind());
+            break;
+        }
+    }
+}
+
+void DeduceTypeVisitor::visit(TypeCastingExpression *expr) {
+    expr->operand()->accept(this);
+    if (!ok()) return;
+
+    EvaluableExprVisitor visitor;
+    expr->operand()->accept(&visitor);
+
+    if (!visitor.ok()) {
+        if (TypeCastingExpression::validateTypeCast(type_, expr->type())) {
+            type_ = expr->type();
+            return;
+        }
+        std::stringstream out;
+        out << "Can not convert " << expr->operand() << " 's type : " << type_ << " to "
+            << expr->type();
+        status_ = Status::SemanticError(out.str());
+        return;
+    }
+    QueryExpressionContext ctx(nullptr);
+    auto val = expr->eval(ctx(nullptr));
+    if (val.isNull()) {
+        status_ =
+            Status::SemanticError("`%s` is not a valid expression ", expr->toString().c_str());
+        return;
+    }
+    type_ = val.type();
+    status_ = Status::OK();
+}
+
+void DeduceTypeVisitor::visit(LabelExpression *expr) {
+    status_ = Status::SemanticError("LabelExpression can not be instantiated: %s",
+                                    expr->toString().c_str());
+}
```

Comment:
Please change the test accordingly at __tests/query/v2/test_label_expr.py:41__

Reply:
sorry for my fault. I will finish local testing successfully in the future PR.

Fixed it.